### PR TITLE
Add more algorithms for ctfe public key

### PR DESCRIPTION
**Pull request by patflynn**
WIP towards #4 .  Throwing up hear to get a sense of this is more or less what you were thinking.

### CODE REVIEW

**Comment by patflynn** [comment]
sorry about the gradlew version noise. I'll pull that out.

**Comment by loosebazooka** [comment]
use `git commit --amend -s` to satisfy DCO on the commits (you might need to squash down first)

File: build.gradle.kts [hunk 1]
```diff
@@ -27,6 +27,7 @@ dependencies {
     implementation("org.conscrypt:conscrypt-openjdk-uber:2.5.2") {
         because("contains library code for all platforms")
     }
+    implementation("org.bouncycastle:bcutil-jdk15on:1.70")
```
Comment (loosebazooka) [inline build.gradle.kts:30]: does this require java15+?

File: src/main/java/dev/sigstore/encryption/Keys.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+import org.bouncycastle.crypto.params.RSAKeyParameters;
+import org.bouncycastle.crypto.util.PublicKeyFactory;
+import org.bouncycastle.util.io.pem.PemObject;
+import org.bouncycastle.util.io.pem.PemReader;
+
+public class Keys {
+
+  public static PublicKey parsePublicKey(byte[] keyBytes)
+      throws InvalidKeySpecException, IOException, NoSuchAlgorithmException {
+    // get public key
```
Comment (loosebazooka) [inline src/main/java/dev/sigstore/encryption/Keys.java:40]: remove old comment

File: build.gradle.kts (hunk 1 above)
Comment (patflynn) [reply build.gradle.kts:30]: I believe it's java 1.5 and up. 
https://www.bouncycastle.org/latest_releases.html

Comment (loosebazooka) [reply build.gradle.kts:30]: lol! okay

**Review by loosebazooka — APPROVED** [review]
need one more spotless apply.

File: build.gradle.kts (hunk 1 above)
Comment (loosebazooka) [reply build.gradle.kts:30]: https://search.maven.org/artifact/org.bouncycastle/bcutil-jdk18on/1.71/jar

File: src/test/java/dev/sigstore/encryption/KeysTest.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
+package dev.sigstore.encryption;
+
+import com.google.common.io.Resources;
+import org.junit.Assert;
```
[the remark this answers is not in the document]
Comment (patflynn) [reply src/test/java/dev/sigstore/encryption/KeysTest.java]: done

File: build.gradle.kts (hunk 1 above)
Comment (patflynn) [reply build.gradle.kts]: done

File: src/main/java/dev/sigstore/fulcio/client/FulcioValidator.java [hunk 4]
```diff
@@ -41,26 +36,13 @@ public class FulcioValidator {
   private final TrustAnchor fulcioRoot;
 
   public static FulcioValidator newFulcioValidator(
-      byte[] fulcioRoot, @Nullable byte[] ctfePublicKey)
+      byte[] fulcioRoot, byte @Nullable [] ctfePublicKey)
```
Comment (loosebazooka) [inline src/main/java/dev/sigstore/fulcio/client/FulcioValidator.java:39]: this is real weird, spotless did this?

**Review by loosebazooka — APPROVED** [review]

**Comment by patflynn** [comment]
image

File: src/main/java/dev/sigstore/fulcio/client/FulcioValidator.java (hunk 4 above)
Comment (patflynn) [reply src/main/java/dev/sigstore/fulcio/client/FulcioValidator.java:39]: nah that was IntelliJ. blows my mind too. but apparently you shouldn't put nullable in front of a primitive type

**Comment by patflynn** [comment]
I've sent a new commit that registers the bouncycastle provider at runtime. I'm 99% sure we shouldn't do this, but I'm curious to see what happens.

File: src/test/java/dev/sigstore/encryption/KeysTest.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+  @Test
+  void parsePublicKey_ec() throws IOException, InvalidKeySpecException, NoSuchAlgorithmException {
+    PublicKey result =
+        Keys.parsePublicKey(Resources.toByteArray(Resources.getResource(EC_PUB_PATH)));
+    assertEquals(result.getAlgorithm(), "EC");
+  }
+
+  @Test
+  void parsePublicKey_ed25519()
```
Comment (loosebazooka) [inline src/test/java/dev/sigstore/encryption/KeysTest.java:50]: I believe the "proper" way to do this woud be to write two tests, one for each side of jdk15, and use the annotation [EnabledForJreRange](https://junit.org/junit5/docs/5.8.1/api/org.junit.jupiter.api/org/junit/jupiter/api/condition/EnabledForJreRange.html)

kinda like
```
@EnabledForJreRange(max = JRE.JAVA_14)
void parsePublicKey_ed25519_withBouncyCastle() {
```

and 
```
@EnabledForJreRange(min = JRE.JAVA_15)
void parsePublicKey_ed25519_withStdLib() {
```

**Review by loosebazooka — APPROVED** [review]

File: src/test/java/dev/sigstore/encryption/KeysTest.java (hunk 5 above)
Comment (patflynn) [reply src/test/java/dev/sigstore/encryption/KeysTest.java]: Nice tip!!  Switched to those.

**Comment by patflynn** [comment]
I'm going to need another LGTM

**Review by loosebazooka — APPROVED** [review]

**Pull request merged**